### DISCUSSION TITLE
Is there a way to use this from C# to draw into a GL or Vulkan area?

### DISCUSSION THREAD

**Discussion by philstopford** [gircore/gir.core#1021] [Q&A]
I'm curious whether this has support for the unified renderer, or GL, or Vulkan areas. I didn't see anything obvious, so wanted to ask.

**Comment by badcel** [comment]
In case you wonder if you can activate the GTK renderers it totally depends on your GTK installation as this project is just a binding for GTK and calls given C functions from C#.

As far as I read the news the new renderers will probably be part of GTK 4.14 which is not yet released.

If you build your own set of GTK development binaries including the new renderers there is a chance that the binaries do not have the same name as the ones from the officially supported sources (gnome SDK, msys2, brew). This results in exceptions that the binary can't be found. So it is recommended to build your own set of GirCore binaries. Details can be found in #963.

Alternatively you can just wait for the official release and the updated nuget packages.

**Comment by philstopford** [comment]
Thanks @badcel . At its simplest, my question is really what does GLArea usage via gircore look like? That's what, I think, would help me the most to understand capabilities here. I'm asking because I have some projects running via Eto.Forms which maps to GTK3; it doesn't seem likely they will support GTK4 and I'm looking at options.
Given a little example, I can start investigating further.

**Reply by badcel** [reply]
It should look like documented here: https://docs.gtk.org/gtk4/class.GLArea.html

The actual api looks like: https://gircore.github.io/api/Gtk.GLArea.html

If you miss something feel free to report a bug as this project is "just" a wrapper around the c api. There is almost no additional API available.

Currently there is no sample but if you have some code for GTK 3 perhaps you add a sample for GTK 4 to GirCore as I think it should be possible to convert: https://github.com/gircore/gir.core/tree/main/src/Samples/Gtk-4.0?

Check out the other samples to get started in general.

The original GTK 3 API is here: https://docs.gtk.org/gtk3/class.GLArea.html

If you like you could check the migration guide, too: https://docs.gtk.org/gtk4/migrating-3to4.html

Perhaps Eto hides some things, then you would need todo the same in your program. But I'm not firm in either Eto nor GlArea.

**Comment by ousnius** [comment]
For a test a while back I used OpenTK to import the native OpenGL functions on the respective platform.
You can call ``GLArea.Context.MakeCurrent();`` and use the "OnRender" event to call into OpenGL yourself.

Of course this isn't using any GTK renderers, it's simply making the GLArea the current GL context for you to write your own renderer.

The ``FuncLoader`` class is taken from GtkSharp: https://github.com/GtkSharp/GtkSharp/blob/develop/Source/Libs/Shared/FuncLoader.cs
The ``GLBindingsContext`` I used for testing is made for Windows (opengl32.dll) but can be changed.

```cs
using Gtk;

public class DrawingArea : GLArea
{
	public DrawingArea()
	{
		CanFocus = true;

		SetRequiredVersion(3, 3);

		HasDepthBuffer = true;
		HasStencilBuffer = true;
	}
}
```
```cs
using Gtk;
using OpenTK;
using OpenTK.Graphics.OpenGL;
using System;

internal class Program
{
    class GLBindingsContext : IBindingsContext
    {
        public IntPtr GetProcAddress(string procName)
        {
            const string libName = "opengl32.dll";
            return FuncLoader.GetProcAddress(FuncLoader.LoadLibrary(libName), procName);
        }
    }

    private static int Main(string[] args)
    {
        var application = Application.New("org.test", Gio.ApplicationFlags.FlagsNone);
        application.OnActivate += (sender, args) =>
        {
            var window = ApplicationWindow.New((Application)sender);

            var drawingArea = new DrawingArea();

            // Left click
            var gesture = GestureClick.New();
            gesture.Button = 1;
            gesture.OnPressed += (sender, args) =>
            {
                drawingArea.QueueRender();
            };
            drawingArea.AddController(gesture);

            window.Child = drawingArea;
            window.Show();

            // Load GL binding
            drawingArea.Context.MakeCurrent();
            GL.LoadBindings(new GLBindingsContext());

            drawingArea.OnRender += OnRender;
        };

        return application.Run();
    }

    private static bool OnRender(GLArea sender, GLArea.RenderSignalArgs args)
    {
        var rnd = new Random();
        GL.ClearColor(rnd.NextSingle(), rnd.NextSingle(), rnd.NextSingle(), rnd.NextSingle());
        GL.Clear(ClearBufferMask.ColorBufferBit | ClearBufferMask.DepthBufferBit);
        return true;
    }
}
```
